tests/test_growing_module.py

```diff
@@ -63,6 +63,112 @@ def test_str(self):
+                allow_growing=True,
+            )
+
+        l1 = GrowingModule(
+            torch.nn.Linear(3, 5, bias=False, device=global_device()),
+            tensor_s_shape=(3, 3),
+            tensor_m_shape=(3, 5),
+            allow_growing=False,
+        )
+
+        self.assertIsInstance(l1, GrowingModule)
+
+        l2 = GrowingModule(
+            torch.nn.Linear(5, 7, bias=False, device=global_device()),
+            tensor_s_shape=(5, 5),
+            tensor_m_shape=(5, 7),
+            allow_growing=True,
+            previous_module=l1,
+        )
+
+        self.assertIsInstance(l2, GrowingModule)
+        self.assertTrue(l2.previous_module is l1)
+
+    def test_delete_update(self):
+        l1 = GrowingModule(
+            torch.nn.Linear(3, 5, bias=False, device=global_device()),
+            tensor_s_shape=(3, 3),
+            tensor_m_shape=(3, 5),
+            allow_growing=False,
+        )
+        l2 = GrowingModule(
+            torch.nn.Linear(5, 7, bias=False, device=global_device()),
+            tensor_s_shape=(5, 5),
+            tensor_m_shape=(5, 7),
+            allow_growing=True,
+            previous_module=l1,
+        )
+
+        def reset(layer, first: bool) -> None:
+            dummy_layer = torch.nn.Identity()
+            layer.extended_output_layer = dummy_layer
+            layer.optimal_delta_layer = dummy_layer
+            if not first:
+                layer.extended_input_layer = dummy_layer
+
+        def reset_all():
+            reset(l1, True)
+            reset(l2, False)
+
+        reset_all()
+        l1.delete_update()
+        self.assertIsInstance(l1.extended_output_layer, torch.nn.Identity)
+        self.assertIsNone(l1.optimal_delta_layer)
+
+        reset_all()
+        with self.assertWarns(UserWarning):
+            l2.delete_update(include_previous=False)
+        self.assertIsNone(l2.extended_input_layer)
+        self.assertIsInstance(l1.extended_output_layer, torch.nn.Identity)
+        self.assertIsNone(l2.optimal_delta_layer)
+        self.assertIsInstance(l2.extended_output_layer, torch.nn.Identity)
+
+        reset_all()
+        l2.delete_update()
+        self.assertIsNone(l2.extended_input_layer)
+        self.assertIsNone(l1.extended_output_layer)
+        self.assertIsNone(l2.optimal_delta_layer)
+        self.assertIsInstance(l2.extended_output_layer, torch.nn.Identity)
+
+        reset_all()
+        l2.delete_update(include_output=True)
+        self.assertIsNone(l2.extended_input_layer)
+        self.assertIsNone(l1.extended_output_layer)
+        self.assertIsNone(l2.optimal_delta_layer)
+        self.assertIsNone(l2.extended_output_layer)
+
+        reset_all()
+        l1.extended_output_layer = None
+        l2.delete_update(include_previous=False)
+        self.assertIsNone(l2.extended_input_layer)
+        self.assertIsNone(l2.optimal_delta_layer)
+        self.assertIsInstance(l2.extended_output_layer, torch.nn.Identity)
```

Comment:
This feels like something roughly equivalent to the lines 127-132.

Reply:
It is not. This trigger the case where the `extended_output_layer` is `None` and therefore there is no warning produced. This is necessary to have full coverage.

Reply:
I misread `l1.extended_output_layer = None` and saw `l1.extended_input_layer = None`. My bad.